### ISSUE TITLE
[rustdoc] Rustdoc should prevent long file names.

### ISSUE THREAD

**Issue by BenTheElder** [rust-lang/rust#34023]
I'm not actually certain if handling absurdly long method names should be a goal for rustdoc, but currently they produce filenames that are too long for the filesystem. I discovered this while working with [opencv-rust](https://github.com/kali/opencv-rust) which auto generates c-wrapper methods for opencv, some of which have very long names. A possible proposed solution was that rustdoc abbreviate filenames, I suggest that a good solution might be splitting the name into 255 char chunks .

Eg the problem file: `/home/benjamin/dev/opencv-rust/target/doc/opencv/sys/fn.cv_calib3d_cv_solvePnPRansac_InputArray_objectPoints_InputArray_imagePoints_InputArray_cameraMatrix_InputArray_distCoeffs_OutputArray_rvec_OutputArray_tvec_bool_useExtrinsicGuess_int_iterationsCount_float_reprojectionError_int_minInliersCount_OutputArray_inliers_int_flags.html`

Would become something like: `/home/benjamin/dev/opencv-rust/target/doc/opencv/sys/fn.cv_calib3d_cv_solvePnPRansac_InputArray_objectPoints_InputArray_imagePoints_InputArray_cameraMatrix_InputArray_distCoeffs_OutputArray_rvec_OutputArray_tvec_bool_useExtrinsicGuess_int_iterationsCount_float_reprojectionErr/or_int_minInliersCount_OutputArray_inliers_int_flags.html`

Previous discussion on the rust subreddit: https://www.reddit.com/r/rust/comments/4m29tk/solution_to_file_name_too_long_for_cargo_doc/

**Comment by BenTheElder** [comment]
Small update: This also appears to be some regression of sorts, after switching back to the `1.8.0` toolchain with rustup I can document the crate as these methods are skipped.

**Comment by GuillaumeGomez** [comment]
> Edit, these methods are marked #[doc(hidden)]. Perhaps then the correct answer is a way to make rustdoc not generate docs for hidden methods?Edit, these methods are marked #[doc(hidden)]. Perhaps then the correct answer is a way to make rustdoc not generate docs for hidden methods?

It isn't supposed to.

**Comment by BenTheElder** [comment]
As far as I can tell, this method is in
/target/debug/build/opencv-<hash>/out/calib3d.extern.rs and marked
#[doc(hidden)].
It is then `include!`-ed in /target/debug/build/opencv-<hash>/out/hub.rs in
a public sys module.

As far as I can tell now, it should be marked hidden and excluded from the
docs, but 1.9.0 and nightly appear to attempt to build docs for these
methods anyhow.

**Comment by BenTheElder** [comment]
I've opened a new issue (https://github.com/rust-lang/rust/issues/34025) to reflect what seems to be the actual problem, but I think that avoiding large file names might also be worth discussing so I will leave this open as well for now.

**Comment by retep998** [comment]
Couldn't we just shorten the name? Come up with some pattern, maybe take inspiration from DOS style short filenames, like shorten it to 100 characters and anything beyond that is replaced with ~1 or a higher number if another file has the same name, or maybe tack on a short hash of the name.

**Comment by ssokolow** [comment]
I'd go with the hash since, as long as it's well-defined (eg. take a prefix of `n - hash_length`characters, then hash the remaining characters using `name of hash`, then append the hash to the prefix), it's possible to re-generate it from the source materials and get the same result if necessary.

**Comment by steveklabnik** [comment]
Triage: no change

**Comment by workingjubilee** [comment]
As of Windows 10 patch 1607 (which happened later in 2016, actually), users can now remove the path length limitations, and patches before that are only supported for Enterprise LTSC.

**Comment by lolbinarycat** [comment]
most linux filesystems still have the limitation of 255 bytes per path segment, and 4096 bytes for the entire path, at least according to `linux/limits.h`.

additionally, it seems that [search engines don't like urls longer than 2000 chars](https://web.archive.org/web/20111110004006/http://www.seomofo.com/experiments/title-and-h1-of-this-post-but-for-the-sake-of-keyword-prominence-stuffing-im-going-to-mention-it-again-using-various-synonyms-stemmed-variations-and-of-coursea-big-fat-prominent-font-size-heres-the-stumper-that-stumped-me-what-is-the-max-number-of-chars-in-a-url-that-google-is-willing-to-crawl-and-index-for-whatever-reason-i-thought-i-had-read-somewhere-that-googles-limit-on-urls-was-255-characters-but-that-turned-out-to-be-wrong-so-maybe-i-just-made-that-number-up-the-best-answer-i-could-find-was-this-quote-from-googles-webmaster-trends-analyst-john-mueller-we-can-certainly-crawl-and-index-urls-over-1000-characters-long-but-that-doesnt-mean-that-its-a-good-practice-the-setup-for-this-experiment-is-going-to-be-pretty-simple-im-going-to-edit-the-permalink-of-this-post-to-be-really-really-long-then-im-going-to-see-if-google-indexes-it-i-might-even-see-if-yahoo-and-bing-index-iteven-though-no-one-really-cares-what-those-assholes-are-doing-url-character-limits-unrelated-to-google-the-question-now-is-how-many-characters-should-i-make-the-url-of-this-post-there-are-a-couple-of-sources-ill-reference-to-help-me-make-this-decision-the-first-is-this-quote-from-the-microsoft-support-pages-microsoft-internet-explorer-has-a-maximum-uniform-resource-locator-url-length-of-2083-characters-internet-explorer-also-has-a-maximum-path-length-of-2048-characters-this-limit-applies-to-both-post-request-and-get-request-urls-the-second-source-ill-cite-is-the-http-11-protocol-which-says-the-http-protocol-does-not-place-any-a-priori-limit-on-the-length-of-a-uri-servers-must-be-able-to-handle-the-uri-of-any-resource-they-serve-and-should-be-able-to-handle-uris-of-unbounded-length-if-they-provide-get-based-forms-that-could-generate-such-uris-a-server-should-return-414-request-uri-too-long-status-if-a-uri-is-longer.html)

**Comment by ssokolow** [comment]
@lolbinarycat `linux/limits.h` is for libc functions that have internal limitations. It does not constrain the filesystems and most Linux filesystems have **no** path length limit.

See, for example, the `getcwd(3)` manpage:

> `getwd()`  does  not `malloc(3)` any memory.  The `buf` argument should be a pointer to an array at least `PATH_MAX` bytes long.  If the length of the absolute pathname of the current working directory, including the terminating null byte, exceeds `PATH_MAX` bytes, `NULL` is returned, and `errno` is set to `ENAMETOOLONG`.  (Note that on some systems, `PATH_MAX` may not be a compile-time constant;  furthermore, its value may depend on the filesystem, see `pathconf(3)`.)  For portability and security reasons, use of `getwd()` is deprecated.

See also:

* [PATH_MAX Is Tricky](https://eklitzke.org/path-max-is-tricky)
* [PATH_MAX simply isn't](https://insanecoding.blogspot.com/2007/11/pathmax-simply-isnt.html)

The TL;DR is that `linux/limits.h` defines a limit for paths handled by *syscalls* but, given that you can always break it by mounting a filesystem with `PATH_MAX`-length paths on a mountpoint deep inside another filesystem, it cannot be absolute, and programs/libraries which want to bypass that limitation will work around that by reimplementing by walking up the chain of ancestors and assembling the full path in userspace.

I can confirm that I can generate test paths on ext4 which are over 5000 characters long, despite `PATH_MAX` being 4096 on my system... which does cause some things to then fail to get the working directory.

**Comment by lolbinarycat** [comment]
true, but you'll still encounter issues if you do the trivial approach, and NAME_MAX is still a bit tricky to get around.

**Comment by ssokolow** [comment]
Certainly. I just think it's important to make it clear that "most linux systems" don't have such a limitation... it's just certain standard library APIs that have the limitation.

**Comment by retep998** [comment]
Even if *most* platforms support it fine, it's still *unreasonable* to have such massive filenames that really don't provide any tangible benefit to the user. About the only reason I can think of to *not* do this is because it would break links to affected doc pages, and I personally think it's still worth doing.

**Comment by lolbinarycat** [comment]
To be fair, you could also argue that having type names that are that long is also unreasonable.
